src/base/associative_cache.hh

```diff
@@ -0,0 +1,338 @@
+#include "mem/cache/replacement_policies/replaceable_entry.hh"
+#include "mem/cache/tags/indexing_policies/base.hh"
+#include "types.hh"
+
+namespace gem5
+{
+
+/** Cache Library.
+ *
+ */
+template <typename Entry>
+class AssociativeCache
+{
+    static_assert(std::is_base_of_v<CacheEntry, Entry>,
+                  "Entry should be derived from CacheEntry");
+  public:
+    typedef replacement_policy::Base BaseReplacementPolicy;
+
+  protected:
+    /** Name of the cache. */
+    std::string cache_name;
+
+    /** The number of entries in the cache. */
+    size_t numEntries;
+
+    /** Associativity of the cache. */
+    size_t _associativity;
+
+    /** The replacement policy of the cache. */
+    BaseReplacementPolicy *replPolicy;
+
+    /** Indexing policy of the cache */
+    BaseIndexingPolicy *indexingPolicy;
+
+    /** Number of sets in the cache. */
+    size_t numSets;
+
+    /** The entries */
+    std::vector<Entry> entries;
+
+  private:
+
+    void initParams(size_t _num_entries, size_t _assoc)
+    {
+        fatal_if(!isPowerOf2(_num_entries), "The number of entries of an "
+                 "AssociativeCache<> must be a power of 2");
+        fatal_if(!isPowerOf2(_assoc), "The associativity of an "
+                 "AssociativeCache<> must be a power of 2");
+        for (auto entry_idx = 0; entry_idx < _num_entries; entry_idx++) {
+            Entry *entry = &entries[entry_idx];
+            indexingPolicy->setEntry(entry, entry_idx);
+            entry->replacementData = replPolicy->instantiateEntry();
+        }
+    }
+
+  public:
+
+    /**
+     * Empty constructor - need to call init() later with all args
+     */
+    AssociativeCache() {}
+
+    /**
+     * Public constructor
+     * @param name Name of the cache
+     * @param num_entries total number of entries of the container, the number
+     *   of sets can be calculated dividing this balue by the 'assoc' value
+     * @param associativity number of elements in each associative set
+     * @param entry_size Size of each entry
+     * @param repl_policy replacement policy
+     * @param indexing_policy indexing policy
+     */
+    AssociativeCache(const char *name, const size_t num_entries,
+                     const size_t associativity,
+                     BaseReplacementPolicy *repl_policy,
+                     BaseIndexingPolicy *indexing_policy,
+                     Entry const &init_val = Entry())
+        : cache_name(name), numEntries(num_entries),
+          _associativity(associativity),
+          replPolicy(repl_policy),
+          indexingPolicy(indexing_policy),
+          numSets(num_entries / associativity),
+          entries(num_entries, init_val)
+    {
+        initParams(numEntries, _associativity);
+    }
+
+    ~AssociativeCache() {
+        entries.clear();
+    }
+
+    /**
+     * Return the name of the cache
+     * @return name of the cache
+     */
+    std::string name() const { return cache_name; }
```

Comment:
Remove (with Named)

Reply:
Added derivation from Named() and removed this.